### PR TITLE
Proof resync

### PR DESCRIPTION
**Pull request by ilyasymbiotic**
### **PR Type**
Enhancement, Tests

___

### **Description**
• **Major refactoring**: Renamed `signature_processor` to `entity_processor` to support both signature and aggregation proof processing
• **New aggregation proof sync functionality**: Added complete P2P synchronization support for aggregation proofs including protobuf messages, gRPC services, and sync runner integration
• **Repository enhancements**: Implemented aggregation proof pending operations with marker-based storage approach and comprehensive database operations
• **Configuration updates**: Added separate epoch configurations for signatures and aggregation proofs with new command-line flags
• **Comprehensive test coverage**: Added extensive tests for aggregation proof processing, sync entities, repository operations, and updated existing tests for the processor rename
• **P2P protocol extension**: Extended gRPC service with `WantAggregationProofs` method and corresponding request/response handling
• **Metrics integration**: Added aggregation proof specific metrics while reusing existing signature metrics infrastructure

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["signature_processor"] --> B["entity_processor"]
  B --> C["ProcessSignature"]
  B --> D["ProcessAggregationProof"]
  E["P2P Sync"] --> F["WantSignatures"]
  E --> G["WantAggregationProofs"]
  H["Repository"] --> I["Signature Operations"]
  H --> J["Aggregation Proof Operations"]
  K["Config"] --> L["SignatureEpochsToSync"]
  K --> M["AggProofEpochsToSync"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Tests</strong></td><td><details><summary>10 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>entity_processor_test.go</strong><dd><code>Rename processor and add aggregation proof tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

core/usecase/entity-processor/entity-processor/entity_processor_test.go

• Renamed package from <code>signature_processor</code> to <code>entity_processor</code><br> • <br>Updated all test function names from <code>TestSignatureProcessor_*</code> to <br><code>TestEntityProcessor_*</code><br> • Changed constructor calls from <br><code>NewSignatureProcessor</code> to <code>NewEntityProcessor</code><br> • Added comprehensive <br>tests for <code>ProcessAggregationProof</code> functionality including success <br>cases, error handling, and integration tests

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-d33a569f6ba36517078115808737e5c83f4244af6ed9f6d125928c93f42f3579">+272/-10</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>sync_provider_test.go</strong><dd><code>Update sync provider tests for entity processor</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/sync-provider/sync_provider_test.go

• Updated import from <code>signature_processor</code> to <code>entity_processor</code><br> • <br>Renamed processor variables from <code>signatureProcessor</code> to <code>entityProcessor</code><br> <br>• Changed constructor calls from <code>NewSignatureProcessor</code> to <br><code>NewEntityProcessor</code><br> • Updated syncer configuration with new aggregation <br>proof related fields

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-06b10f37763d25a55ce29018adbc2fa196a020cce2679d2c0cf35659a34db2a5">+55/-31</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>badger_repository_aggregation_proof_pending_test.go</strong><dd><code>Add comprehensive aggregation proof pending repository tests</code></dd></summary>
<hr>

internal/client/repository/badger/badger_repository_aggregation_proof_pending_test.go

• New test file for aggregation proof pending operations<br> • Tests for <br><code>SaveAggregationProofPending</code>, <code>RemoveAggregationProofPending</code>, and <br><code>GetSignatureRequestsWithoutAggregationProof</code><br> • Comprehensive test <br>coverage including edge cases, pagination, and integration scenarios<br> • <br>Tests for duplicate handling and cross-epoch isolation

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-48b4bbdc97bc4c1889a2446e8c509b5a2d12446791095c0bb6615983dae1f7bd">+280/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>entity_aggregation_proof_sync_test.go</strong><dd><code>Add aggregation proof sync entity tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

core/entity/entity_aggregation_proof_sync_test.go

• New test file for aggregation proof sync entities<br> • Tests for <br><code>WantAggregationProofsRequest</code>, <code>WantAggregationProofsResponse</code>, and <br><code>AggregationProofProcessingStats</code><br> • Integration test demonstrating <br>complete aggregation proof sync flow<br> • Comparison tests with signature <br>processing stats to ensure consistency

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-f81f92d9f9bc22b3d610d1d2f503f7515f2c66371b8f81d98320f52df18c82f7">+253/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>signer_app.go</strong><dd><code>Update signer app mocks for entity processor</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/signer-app/mocks/signer_app.go

• Renamed mock interface from <code>MocksignatureProcessor</code> to <br><code>MockentityProcessor</code><br> • Added <code>ProcessAggregationProof</code> method to entity <br>processor mock<br> • Updated all method signatures and recorder names to <br>match new interface<br> • Maintained existing <code>ProcessSignature</code> mock <br>functionality

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-003d3d91b36347f5f0eeadb7f09c29dcf4f12297c926dcea9ed3e3cdbf24ee56">+28/-14</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>signature_listener_uc.go</strong><dd><code>Update signature listener mocks for entity processor</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/signature-listener/mocks/signature_listener_uc.go

• Renamed mock interface from <code>MocksignatureProcessor</code> to <br><code>MockentityProcessor</code><br> • Updated all mock method names and recorder types<br> <br>• Maintained existing <code>ProcessSignature</code> mock functionality<br> • Updated <br>import references and type names throughout

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-d91700762a254499e5c699d751e9294125fd0b7d67c4b652a5e99b16d596cf82">+14/-14</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>p2p_grpc_send_want_signatures_request_test.go</strong><dd><code>Update gRPC tests for dual handler support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/client/p2p/p2p_grpc_send_want_signatures_request_test.go

• Updated test setup to use new <code>GRPCHandler</code> constructor with separate <br>handlers<br> • Added mock aggregation proof request handler for test <br>compatibility<br> • Updated handler initialization calls throughout test <br>cases<br> • Maintained existing signature request test functionality

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-f0dd73d3ab00682d98c0ab133bb2a01284446befd37d5fc209138f88168d896d">+15/-2</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>signer_app_test.go</strong><dd><code>Update signer app tests for entity processor</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/signer-app/signer_app_test.go

• Updated imports from <code>signature_processor</code> to <code>entity_processor</code><br> • <br>Changed processor creation from <code>NewSignatureProcessor</code> to <br><code>NewEntityProcessor</code><br> • Updated configuration field name from <br><code>SignatureProcessor</code> to <code>EntityProcessor</code><br> • Maintained existing test <br>functionality with new processor interface

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-55eb4622bf16d7eb4d6b23fbea6b071a3f232d2a023cb0f73f4bf98e32bf0692">+9/-9</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>badger_repository_signature_request_test.go</strong><dd><code>Add duplicate handling tests for pending signature requests</code></dd></summary>
<hr>

internal/client/repository/badger/badger_repository_signature_request_test.go

• Added test for duplicate handling in <code>SaveSignatureRequestPending</code><br> • <br>Tests error scenarios when trying to save same pending request twice<br> • <br>Verifies proper error types and messages for duplicate operations<br> • <br>Ensures existing pending requests remain functional after duplicate <br>attempts

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-154f704a125e2b4028cfe52e25da864300c159a11137559bafb0f7487a309f07">+28/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>p2p_grpc_test.go</strong><dd><code>Rename handler field to signatureHandler in tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/client/p2p/p2p_grpc_test.go

• Renamed struct field from <code>handler</code> to <code>signatureHandler</code> in test setup

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-e1cdbcb69a3da51788f036f6436fb1732f8ad543a31f82cf5ed2a41ad2b37bbb">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>18 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>sync.pb.go</strong><dd><code>Add aggregation proof protobuf messages and service</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/client/p2p/proto/v1/sync.pb.go

• Added new protobuf message types <code>WantAggregationProofsRequest</code> and <br><code>WantAggregationProofsResponse</code><br> • Added new gRPC service method <br><code>WantAggregationProofs</code> to <code>SymbioticP2PService</code><br> • Updated protobuf <br>imports to include <code>v1/message.proto</code><br> • Extended message type arrays and <br>dependency indices for new aggregation proof messages

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-abe72a803d4399a96cf049a58056a1594b1637fe57ad3016d856978670c8b684">+131/-24</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>sync_grpc.pb.go</strong><dd><code>Add aggregation proof gRPC service methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/client/p2p/proto/v1/sync_grpc.pb.go

• Added <code>WantAggregationProofs</code> method to gRPC client and server <br>interfaces<br> • Updated service descriptor with new aggregation proof <br>method handler<br> • Added unimplemented method stub for aggregation proof <br>requests<br> • Extended full method name constants for new service method

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-268b8b8dd704d0d28316b5b549a5ed4e35e515b39b96a80b06ec791603452e36">+39/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>badger_repository_aggregation_proof.go</strong><dd><code>Add aggregation proof pending repository operations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/client/repository/badger/badger_repository_aggregation_proof.go

• Added <code>SaveAggregationProofPending</code>, <code>RemoveAggregationProofPending</code> <br>methods for pending aggregation proof management<br> • Implemented <br><code>GetSignatureRequestsWithoutAggregationProof</code> with pagination support<br> • <br>Added key generation functions for aggregation proof pending <br>operations<br> • Enhanced error handling with proper entity not found <br>checks

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-da59e69cb72ea65614d7d9573ac77efcfd10c2b296f96bceba75d7187cb6509b">+131/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>sync_runner.go</strong><dd><code>Add aggregation proof sync runner functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/sync-runner/sync_runner.go

• Added aggregation proof sync interfaces and methods to <code>p2pService</code> <br>and <code>provider</code><br> • Implemented <code>runAggregationProofSync</code> method alongside <br>existing signature sync<br> • Updated metrics interface with aggregation <br>proof specific methods<br> • Modified main sync loop to run both signature <br>and aggregation proof syncs independently

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-29c37899dd5cdd31774d651f688ff77a831e7af05ebac0c18a5f2277f18d8f69">+69/-6</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>p2p_grpc_send_want_aggregation_proofs_request.go</strong><dd><code>Add aggregation proof gRPC client implementation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/client/p2p/p2p_grpc_send_want_aggregation_proofs_request.go

• New file implementing aggregation proof request/response handling <br>over gRPC<br> • Added <code>SendWantAggregationProofsRequest</code> method for peer <br>communication<br> • Implemented protobuf conversion functions for <br>aggregation proof entities<br> • Added timeout handling and error <br>management for aggregation proof requests

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-f8d954f390644c9ed2aa269a4968f298c3fa3b5932a058c851d544e5fef2fa86">+148/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>badger_repository_signature_request.go</strong><dd><code>Refactor signature request pending storage to marker-based approach</code></dd></summary>
<hr>

internal/client/repository/badger/badger_repository_signature_request.go

• Modified <code>SaveSignatureRequestPending</code> to store only markers instead <br>of full data<br> • Refactored <code>GetSignatureRequestsByEpochPending</code> to use <br>marker-based approach with hash index lookups<br> • Added duplicate check <br>for pending signature requests<br> • Enhanced error handling for missing <br>signature requests in pending collection

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-2021ea2acd66040976cb26ff27f35e109e0037a4e3df435c8a7487f84adf71cb">+104/-12</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>root.go</strong><dd><code>Update main application to use entity processor</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

cmd/relay/root.go

• Updated import from <code>signature_processor</code> to <code>entity_processor</code><br> • <br>Renamed processor variables and constructor calls throughout<br> • Added <br>aggregation proof configuration fields to sync provider<br> • Updated <br>signer app and signature listener configurations with new processor <br>interface

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-c1698ba36d063bfa002e73dd2d0a518c17e10cbcb5b4895f3a9b31864bc9dbdb">+17/-14</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>entity_processor.go</strong><dd><code>Rename processor and add aggregation proof processing</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

core/usecase/entity-processor/entity-processor/entity_processor.go

• Renamed package and struct from <code>signature_processor</code> to <br><code>entity_processor</code><br> • Added aggregation proof repository methods to <br>interface<br> • Implemented <code>ProcessAggregationProof</code> method for handling <br>aggregation proof messages<br> • Enhanced <code>ProcessSignature</code> to save <br>aggregation proof pending markers for aggregation keys

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-8de17be6e3cd4c32e8827bd2cbf256c050082bea5090dde44065a692fe5a503a">+35/-7</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>p2p_grpc_handler.go</strong><dd><code>Add aggregation proof gRPC handler support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/client/p2p/p2p_grpc_handler.go

• Added <code>aggregationProofRequestHandler</code> interface for handling <br>aggregation proof requests<br> • Updated <code>GRPCHandler</code> to accept both <br>signature and aggregation proof handlers<br> • Implemented <br><code>WantAggregationProofs</code> gRPC method handler<br> • Separated concerns between <br>signature and aggregation proof handling

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-504fa49b050f552656baaf63c1842986b84e6f5a98d53bd6efa38d258d04b8ca">+28/-4</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>sync_provider_build_want_aggregation_proofs_request.go</strong><dd><code>Add aggregation proof request building functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/sync-provider/sync_provider_build_want_aggregation_proofs_request.go

• New file implementing aggregation proof request building logic<br> • <br>Added <code>BuildWantAggregationProofsRequest</code> method with epoch-based <br>scanning<br> • Implemented pagination support for large numbers of pending <br>aggregation proofs<br> • Added prioritization of recent epochs over older <br>ones

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-22dd43884e670009be1bc8743b194c72d49d461baae4491f83a4386e8f7a478f">+68/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>sync_provider_process_received_aggregation_proofs.go</strong><dd><code>Add aggregation proof processing functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/sync-provider/sync_provider_process_received_aggregation_proofs.go

• New file implementing aggregation proof processing logic<br> • Added <br><code>ProcessReceivedAggregationProofs</code> method with comprehensive error <br>handling<br> • Implemented statistics tracking for various processing <br>outcomes<br> • Added validation of signature requests before processing <br>aggregation proofs

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-afb9eabcd49aeed3d5e80d9dc5293e88bd0dae21bdd7ef5a23e31385f7ac3f67">+57/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>signer_app.go</strong><dd><code>Update signer app to use entity processor interface</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/signer-app/signer_app.go

• Renamed interface from <code>signatureProcessor</code> to <code>entityProcessor</code><br> • Added <br><code>ProcessAggregationProof</code> method to processor interface<br> • Updated <br>configuration field from <code>SignatureProcessor</code> to <code>EntityProcessor</code><br> • <br>Updated method calls to use new processor interface

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-0544b8fb7016065ff2de8522f81490696d1e5b79cd86efa32d652a7b865bff41">+10/-9</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>sync_provider.go</strong><dd><code>Update sync provider interfaces for aggregation proof support</code></dd></summary>
<hr>

internal/usecase/sync-provider/sync_provider.go

• Added aggregation proof repository methods to <code>repo</code> interface<br> • <br>Renamed <code>signatureProcessor</code> to <code>entityProcessor</code> with additional methods<br> <br>• Added aggregation proof configuration fields to <code>Config</code> struct<br> • <br>Updated validation and interface definitions for aggregation proof <br>support

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-e42fd4485424fee0318caa5c5456a419ddc4e32f66d2a7c848e846c96532bcef">+9/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>signature_listener_uc.go</strong><dd><code>Update signature listener to use entity processor</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/signature-listener/signature_listener_uc.go

• Renamed interface from <code>signatureProcessor</code> to <code>entityProcessor</code><br> • <br>Updated configuration field from <code>SignatureProcessor</code> to <code>EntityProcessor</code><br> <br>• Changed method calls to use new processor interface<br> • Maintained <br>existing signature processing functionality

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-49a39689b9881e49aeed3e5efe09cce911ae9dd59d215f08a7a1850470f3c765">+6/-6</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>entity_signature_sync.go</strong><dd><code>Add aggregation proof sync entity definitions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

core/entity/entity_signature_sync.go

• Added <code>WantAggregationProofsRequest</code> and <code>WantAggregationProofsResponse</code> <br>entity types<br> • Implemented <code>AggregationProofProcessingStats</code> with error <br>counting functionality<br> • Added <code>TotalErrors</code> method for aggregation <br>proof statistics<br> • Extended signature sync entities with aggregation <br>proof support

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-c29076a0b890cc04d5ba494c5d50f82265956714ecdfd4d06c7275c01cd8ce7c">+28/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>sync_provider_handle_want_aggregation_proofs_request.go</strong><dd><code>Add aggregation proof request handling functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/sync-provider/sync_provider_handle_want_aggregation_proofs_request.go

• New file implementing aggregation proof request handling<br> • Added <br><code>HandleWantAggregationProofsRequest</code> method for peer requests<br> • <br>Implemented response count limiting and error handling<br> • Added logic <br>to skip missing aggregation proofs gracefully

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-87c73fb5e9e79d3204c4a6c0abc798eea42974d4a414d7594a41502d189c207a">+42/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>metrics.go</strong><dd><code>Add aggregation proof metrics support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/metrics/metrics.go

• Added <code>ObserveP2PSyncAggregationProofsProcessed</code> method for <br>aggregation proof metrics<br> • Added <br><code>ObserveP2PSyncRequestedAggregationProofs</code> method for request counting<br> • <br>Reused existing signature metrics infrastructure for aggregation <br>proofs<br> • Added comments indicating future separation of metrics if <br>needed

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-bd9e2b45f2d8bd49e5cddad0b05f5971fd036df143c86aa882012adccb090744">+12/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>sync.proto</strong><dd><code>Add aggregation proof synchronization to P2P protocol</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/client/p2p/proto/v1/sync.proto

• Added import for <code>v1/message.proto</code><br> • Added new <code>WantAggregationProofs</code> <br>RPC method to service<br> • Defined <code>WantAggregationProofsRequest</code> and <br><code>WantAggregationProofsResponse</code> message types

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-372e03e5e700cef059e3431a80ff63592ae55c2ee400c91badc8249fe173cdce">+13/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>config.go</strong><dd><code>Add aggregation proof sync configuration options</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

cmd/relay/config.go

• Updated <code>SyncConfig</code> struct with separate epoch configurations for <br>signatures and aggregation proofs<br> • Added new command line flags for <br><code>signature-epochs</code> and <code>agg-proof-epochs</code><br> • Updated flag binding and <br>configuration parsing for new aggregation proof settings<br> • Maintained <br>backward compatibility while adding new configuration options

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-e8c6ba868ebf52e360af9fd3ec6340d1015f8644c58cfc3415f1f63394591eb6">+11/-6</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Refactoring</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>signature_listener_uc_test.go</strong><dd><code>Refactor signature processor to entity processor in tests</code></dd></summary>
<hr>

internal/usecase/signature-listener/signature_listener_uc_test.go

• Updated import from <code>signature_processor</code> to <code>entity_processor</code><br> • <br>Changed processor initialization from <code>NewSignatureProcessor</code> to <br><code>NewEntityProcessor</code><br> • Renamed config field from <code>SignatureProcessor</code> to <br><code>EntityProcessor</code>

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-7a883c583e5937c16e0e8e840ff4c11e480833fd3d1899b8f36178ec4351fd88">+4/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>signer_app_handle_signature_aggregated_message.go</strong><dd><code>Delegate aggregation proof handling to entity processor</code>&nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/signer-app/signer_app_handle_signature_aggregated_message.go

• Replaced direct repository call <code>SaveAggregationProof</code> with processor <br>method <code>ProcessAggregationProof</code><br> • Changed from passing individual <br>parameters to passing the entire message object

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-0a29a53785d7f13606e803f53d435b2e8cd8301b1e4e69d3e2d18b092a20f013">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>sync_provider_build_want_signatures_map.go</strong><dd><code>Rename epochs configuration field for signature syncing</code>&nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/sync-provider/sync_provider_build_want_signatures_map.go

• Renamed configuration field from <code>EpochsToSync</code> to <br><code>SignatureEpochsToSync</code><br> • Updated all references to use the new field <br>name

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-0bfcf3af953f2c9d44ca765ea5ca74ffcf6c02e64b4e28d5033b3da554588b2e">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>sync_provider_process_received_signatures.go</strong><dd><code>Switch to entity processor for signature processing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/usecase/sync-provider/sync_provider_process_received_signatures.go

• Changed from using <code>SignatureProcessor</code> to <code>EntityProcessor</code> for <br>processing signatures<br> • Updated method call to use the new processor <br>interface

</details>

  </td>
  <td><a href="https://github.com/symbioticfi/relay/pull/172/files#diff-764953cb75706cccadfda0a431aa981b468daeeea3ffc5f1c36a5a4865455c35">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: cmd/relay/config.go [hunk 1]
```diff
@@ -135,10 +135,11 @@ type CacheConfig struct {
...
 type SyncConfig struct {
-	Enabled      bool          `mapstructure:"enabled"`
-	Period       time.Duration `mapstructure:"period"`
-	Timeout      time.Duration `mapstructure:"timeout"`
-	EpochsToSync uint64        `mapstructure:"epochs"`
+	Enabled               bool          `mapstructure:"enabled"`
+	Period                time.Duration `mapstructure:"period"`
+	Timeout               time.Duration `mapstructure:"timeout"`
+	SignatureEpochsToSync uint64        `mapstructure:"signature-epochs"`
+	AggProofEpochsToSync  uint64        `mapstructure:"agg-proof-epochs"`
```
[the remark this answers is not in the document]
Comment (algys) [reply cmd/relay/config.go:142]: yeah, I don't think we need 2 separate configs for this

File: core/usecase/entity-processor/entity-processor/entity_processor.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+	}, nil
+}
+
+// ProcessSignature processes a signature with SignatureMap operations and optionally saves SignatureRequest
+func (s *EntityProcessor) ProcessSignature(ctx context.Context, param entity.SaveSignatureParam) error {
+	publicKey, err := crypto.NewPublicKey(param.KeyTag.Type(), param.Signature.PublicKey)
+	if err != nil {
+		return errors.Errorf("failed to get public key: %w", err)
+	}
+	err = publicKey.VerifyWithHash(param.Signature.MessageHash, param.Signature.Signature)
```
Comment (algys) [inline core/usecase/entity-processor/entity-processor/entity_processor.go:82]: how do you verify that `param.Signature` is related to `param.RequestHash` ?
It looks like I can pass any random valid signature here even if it's not related to `param.RequestHash`

File: core/usecase/entity-processor/entity-processor/entity_processor.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+}
+
+// ProcessAggregationProof processes an aggregation proof message by saving it and removing from pending collection
+func (s *EntityProcessor) ProcessAggregationProof(ctx context.Context, msg entity.AggregatedSignatureMessage) error {
+	validatorSet, err := s.cfg.Repo.GetValidatorSetByEpoch(ctx, uint64(msg.Epoch))
+	if err != nil {
+		return errors.Errorf("failed to get validator set: %w", err)
+	}
+
+	ok, err := s.cfg.Aggregator.Verify(validatorSet, msg.KeyTag, msg.AggregationProof)
```
Comment (algys) [inline core/usecase/entity-processor/entity-processor/entity_processor.go:185]: same question here how proof relates to requestHash

**Review by algys — CHANGES_REQUESTED** [review]
